af/analysis/ooni_db_backup.py

```diff
@@ -0,0 +1,287 @@
+log = logging.getLogger("ooni-db-backup")
+
+# TODO backup schema
+
+# backup implemented in release 22.10, 2022-10-25
+sql_base = """
+    BACKUP TABLE %(tblname)s TO
+    S3(%(s3path_base)s, %(aws_id)s, %(aws_secret)s)
+    SETTINGS compression_method='lzma', compression_level=3
+"""
+sql_incremental = """
+    BACKUP TABLE %(tblname)s TO
+    S3(%(s3path_incremental)s, %(aws_id)s, %(aws_secret)s)
+    SETTINGS
+    base_backup = S3(%(s3path_base)s, %(aws_id)s, %(aws_secret)s)
+    compression_method='lzma', compression_level=3
+"""
+
+
+@metrics.timer("run_backup")
+def run_backup(conf) -> None:
+    aws_id = conf["public_aws_access_key_id"]
+    aws_secret = conf["public_aws_secret_access_key"]
+    bucket_name = conf["public_bucket_name"]
+    baseurl = f"https://{bucket_name}.s3.amazonaws.com/clickhouse_backup/"
+    click = Clickhouse.from_url(conf["clickhouse_url"])
+
+    for tblname, tbl_setting in conf["tables"].items():
+        if tbl_setting == "ignore":
+            log.info(f"Skipping {tblname}")
+            continue
+        elif tbl_setting == "full":
+            sql = sql_base
+        elif tbl_setting == "incremental":
+            sql = sql_incremental
+        else:
+            log.error(f"Skipping {tblname} unknown value {tbl_setting}")
+            continue
+
+        log.info(f"backing up {tblname}")
+        t0 = perf_counter()
+
+        s3path_base = f"{baseurl}/{tblname}__base.bkp"
+        s3path_incremental = f"{baseurl}/{tblname}__incremental.bkp"
+
+        kw = dict(
+            aws_id=aws_id,
+            aws_secret=aws_secret,
+            s3path_base=s3path_base,
+            s3path_incremental=s3path_incremental,
+            tblname=tblname,
+        )
+        click.execute(sql, kw)
+
+        delta_ms = int((perf_counter() - t0) * 1000)
+        metrics.timing(f"table_{tblname}", delta_ms)
+        log.info("table backup completed")
+
+
+def describe_table(click, tblname):
+    assert " " not in tblname
+    assert "'" not in tblname
+    colnames = []
+    blob = []
+    for row in click.execute(f"DESCRIBE TABLE {tblname}"):
+        colname, coltype = row[:2]
+        colnames.append(colname)
+        blob.append(f"{colname} {coltype}")
+
+    assert colnames
+    s_blob = ", ".join(blob)  # "measurement_uid String report_id String, ... "
+    s_colnames = ", ".join(colnames)  # "measurement_uid, report_id, ... "
+    return s_colnames, s_blob
+
+
+def export_fastpath(click, aws_id: str, aws_secret: str, baseurl: str) -> None:
+    tblname = "fastpath"
+    s_colnames, s_blob = describe_table(click, tblname)
+    assert s_blob.startswith("measurement_uid String, ")
+    # s_colnames is a string to be placed into ##colnames## without quotes
+    basesql = """
+    INSERT INTO FUNCTION s3(%(s3path)s, %(aws_id)s, %(aws_secret)s,
+        'CSVWithNames',
+        %(s_blob)s
+    )
+    SELECT ##colnames##
+    FROM fastpath
+    WHERE toYYYYMM(measurement_start_time) = %(yyyymm)s
```

Comment:
I think the tricky thing here is that people can upload old measurements later, so if at a certain time you have backed up data from a given month, you will have to back it up again every time you run the backup process because you don't have the guarantee that other measurements have been uploaded in the meantime.

How about instead you use the report_id or measurement_uid prefixes for determining the partitions of the csv backup files?

If you decide to keep it like this, you may still want to do the partitioning into files at the clickhouse level instead of wrapping it in a for loop, like this: https://clickhouse.com/docs/en/integrations/s3/s3-table-functions/#splitting-large-files.

Reply:
> if at a certain time you have backed up data from a given month,

The initial version of the tool is not designed to do incremental backups. When we switch to a newer version of Clickhouse we can use its native, incremental backup support.

> How about instead you use the report_id or measurement_uid prefixes for determining the partitions of the csv backup files?

This is a backup tool meant to run against our only production database. The table is indexed by measurement_start_time and we have to follow the index if we want to minimize the CPU and I/O load on the database.
Additionally, partitioning the files by YYYYMM makes it easier to understand the contents and reuse them.
Also, native partitioning by clickhouse does not seem to be supported by the current version.

Reply:
Ah that's unfortunate that partioned writes are not supported in the version of clickhouse we are using.

According to the changelog support for it was introduced in v21.10 (https://clickhouse.com/docs/en/whats-new/changelog/2021/#clickhouse-release-v2110-2021-10-16).

> Additionally, partitioning the files by YYYYMM makes it easier to understand the contents and reuse them.

My suggestion was about using the prefix from the report_id or measurement_id, so you would effectively end up with something which is the same as using the `measurement_start_time`, except that the content of a particular file will not change over time.

I do understand that there are load constraints to take into account and I guess it's OK for the time being to use this approach.